src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,186 @@
 package core.basesyntax;
 
+import java.util.Collection;
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
+    public MyLinkedList(Collection<? extends T> col) {
+        addAll((List<T>) col);
+    }
+
+    public MyLinkedList() {
+    }
+
     @Override
     public void add(T value) {
+        if (isFirst()) {
+            addFirst(value);
+            return;
+        }
+        addLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index);
+        if (index == 0) {
+            addFirst(value);
+        } else if (index == size) {
+            addLast(value);
+        } else {
+            addInMiddle(value, findNode(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
+    }
+
+    private void addFirst(T value) {
+        Node<T> f = first;
+        Node<T> newNode = new Node<>(null, value, f);
+        first = newNode;
+        if (f == null) {
+            last = newNode;
+        } else {
+            f.prev = newNode;
+        }
+        size++;
+    }
+
+    private void addLast(T value) {
+        final Node<T> l = last;
+        Node<T> newNode = new Node<T>(l, value, null);
+        last = newNode;
+        l.next = newNode;
+        size++;
+    }
+
+    private void addInMiddle(T value, Node<T> successor) {
+        Node<T> predecessor = successor.prev;
+        Node<T> newNode = new Node<>(predecessor, value, successor);
+        successor.prev = newNode;
+        if (predecessor == null) {
+            first = newNode;
+        } else {
+            predecessor.next = newNode;
+        }
```

Comment:
pay attention to the order of data assignment, because in this variant you simply overwrote the object before checking
```suggestion
        if (predecessor == null) {
            first = newNode;
        } else {
            predecessor.next = newNode;
        }
        successor.prev = newNode;
```